### PR TITLE
Test linalg vector norm subgradient and fix nans when p < 1

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74205
* __->__ #75103

This PR:
- ignores sanitizer for pow kernel when p = -2 (ignore does not apply recursively, so need to apply at the root)
- splits OpInfo so that one tests zeros as inputs (with a single OpInfo we'd need to skip the gradgrad tests)
- corrects a division by zero issue that produces NaNs when p < 1
- avoids an extra unneeded masked fill for the inf norm case

### CODE REVIEW

**Comment by soulitzer** [comment]
@albanD looks like ASAN tests are already failing for backward already when zero is passed, I think it is ok to just silence it

**Comment by albanD** [comment]
Yes. It should be a high pri issue to fix it though.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -190,6 +190,8 @@ Tensor norm_backward(const Tensor& grad, const Tensor& self, const optional<Scal
 }
 
 Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p_, Tensor norm, IntArrayRef dim, bool keepdim) {
+  // NB: zero entries of `norm` (and in one case, `self`) are replaced with something
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:193]: What does "one case" means here? Does that refer to the test cases?

**Review by albanD — COMMENTED** [review]
cc @ngimel what is your feeling about this? Is the perf hit worth it to appease ASAN build?

**Comment by ngimel** [comment]
I don't think masked_fill just to appease ASAN is worth it, we are filling those values later anyway with zeros, right?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -189,7 +189,10 @@ Tensor norm_backward(const Tensor& grad, const Tensor& self, const optional<Scal
   return norm_backward(grad, self, p_, norm, {}, true);
 }
 
-Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p_, Tensor norm, IntArrayRef dim, bool keepdim) {
+Tensor norm_backward(
+    Tensor grad, const Tensor& self, const optional<Scalar> & p_, Tensor norm, IntArrayRef dim, bool keepdim) {
+  // NB: we could've masked_fill the zero elements with an arbitrary value like 1 to
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:194]: I don't think this comment is precise enough. What are you maske filling? And why?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -205,31 +208,35 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
...
+    return self_scaled * grad * norm.pow(1 - p);
   } else if (p < 2.0) {
-    self_scaled = self.sgn() * self.abs().pow(p - 1);
+    self_scaled = self.sgn() * self.abs().pow_(p - 1);
     scale_v = grad / norm.pow(p - 1);
+    scale_v.masked_fill_(norm == 0, 0);
+    return self_scaled * scale_v;
   } else {
-    self_scaled = self * self.abs().pow(p - 2);
+    self_scaled = self * self.abs().pow_(p - 2);
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:235]: If we're adding inplace, we can make the multiplication inplace as well here.
Here and above

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 4]
```diff
@@ -205,31 +208,35 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
...
     scale_v = grad / norm.pow(p - 1);
+    scale_v.masked_fill_(norm == 0, 0);
+    return self_scaled * scale_v;
   }
-  // handle case at 0 where we return a subgradient containing 0
-  scale_v.masked_fill_(norm == 0, 0);
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:231]: If there is only one case where you don't do it, then I think it is worth leaving this here and having the one that doesn't need this do an early return?

**Review by albanD — COMMENTED** [review]
Do we have an issue tracking what needs to done

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 4 above)
Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.cpp:231]: Hmm there are two actually, the p < 1.0 case masks self and doesn't need to mask norm, so there will a little extra logic to handle that if I move it to below

Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:231]: Ok then!

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.cpp]: Updated!

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 5]
```diff
@@ -189,7 +189,12 @@ Tensor norm_backward(const Tensor& grad, const Tensor& self, const optional<Scal
   return norm_backward(grad, self, p_, norm, {}, true);
 }
 
-Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p_, Tensor norm, IntArrayRef dim, bool keepdim) {
+Tensor norm_backward(
+    Tensor grad, const Tensor& self, const optional<Scalar> & p_, Tensor norm, IntArrayRef dim, bool keepdim) {
+  // NB: We mask fill the NaNs in the output to be zero but still do float division
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:194]: nit: this should be past tense no?

**Review by albanD — APPROVED** [review]
Details on the comment. But good otherwise!

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 5 above)
Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.cpp:194]: hmm... debatable :)

**Pull request closed without merging**